File: day15/src/main.py
#!/usr/bin/env pyton

from queue import PriorityQueue


class Cell:
    def __init__(self, i, j, risk, score, parent):
        self.i = i
        self.j = j
        self.risk = risk
        self.score = score
        self.parent = parent
        self.childs = []

    def __lt__(self, other):
        return (self.i, self.j) < (other.i, other.j)


def dijkstra(graph, start, end):
    start.score = 0
    queue = PriorityQueue()
    for node in graph:
        if node != start:
            node.score = float("inf")
            node.parent = None
        queue.put((node.score, node))
    while not queue.empty():
        (prio, cell) = queue.get()
        for child in cell.childs:
            score = cell.score + child.risk
            if score < child.score:
                child.score = score
                child.parent = cell
                child = queue.put((child.score, child))
    return end.score


def main():
    #
    # Part 1
    #

    # Build map
    map = []
    with open("./data/input.txt") as f:
        for i, line in enumerate(f.readlines()):
            row = []
            for j, risk in enumerate(line.strip()):
                cell = Cell(i, j, int(risk), None, None)
                row.append(cell)
            map.append(row)
    # Update nodes with their childs
    for i in range(len(map)):
        for j in range(len(map[0])):
            for (ii, jj) in [(i - 1, j), (i, j - 1), (i, j + 1), (i + 1, j)]:
                if 0 <= ii < len(map) and 0 <= jj < len(map[0]):
                    map[i][j].childs.append(map[ii][jj])
    # Shortest path
    graph = [cell for row in map for cell in row]
    score = dijkstra(graph, map[0][0], map[-1][-1])
    print("Part 1 - shortest=", score)

    #
    # Part 2
    #

    # Build new map
    map5 = []
    for mi in range(5):
        for i in range(len(map)):
            row = []
            map5.append(row)
            for mj in range(5):
                for j in range(len(map[0])):
                    ii = mi * len(map) + i
                    jj = mj * len(map[0]) + j
                    cell = Cell(
                        ii, jj, (map[i][j].risk - 1 + mi + mj) % 9 + 1, None, None
                    )
                    row.append(cell)
    # Update nodes with their childs
    for i in range(len(map5)):
        for j in range(len(map5[0])):
            for (ii, jj) in [(i - 1, j), (i, j - 1), (i, j + 1), (i + 1, j)]:
                if 0 <= ii < len(map5) and 0 <= jj < len(map5[0]):
                    map5[i][j].childs.append(map5[ii][jj])
    # Find shortest path
    graph5 = [cell for row in map5 for cell in row]
    score = dijkstra(graph5, map5[0][0], map5[-1][-1])
    print("Part 2 - shortest=", score)


if __name__ == "__main__":
    main()

File: day15/src/main.rs
// https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm

use std::error::Error;
use std::fs;
use std::cell::RefCell;
use std::rc::Rc;
// use std::cmp::Eq;
use priority_queue::PriorityQueue;

#[derive(Clone, Hash, PartialEq)]
struct Cell {
    i: usize,
    j: usize,
    risk: usize,
    score: usize,
    parent: Option<Rc<RefCell<Cell>>>,
    childs: Vec<Rc<RefCell<Cell>>>,
}

impl Cell {
    fn new(i: usize, j: usize, risk: usize) -> Cell {
        Cell {
            i,
            j,
            risk,
            score: 0,
            parent: None,
            childs: Vec::new()
        }
    }
}

fn display_board(map: &Vec<Vec<Cell>>) {
    let mut out = String::new();
    for line in map {
        for cell in line {
            out += &format!("{:2},", cell.score);
        }
        out += "\n";
    }
    println!("{}", out);
}

fn djikstra(
    graph: &mut Vec<Rc<RefCell<Cell>>>, 
    start: Rc<RefCell<Cell>>, 
    end: Rc<RefCell<Cell>>,
    path: &Vec<Rc<RefCell<Cell>>>
) -> usize {

    // dist[source] ← 0                           // Initialization
    start.borrow().score = 0;

    // create vertex priority queue Q
    let mut prioqueue = PriorityQueue::new();

    // for each vertex v in Graph:          
    //     if v ≠ source
    //         dist[v] ← INFINITY                 // Unknown distance from source to v
    //         prev[v] ← UNDEFINED                // Predecessor of v
    //     Q.add_with_priority(v, dist[v])    
    for vertex in graph {
        if vertex != start {
            vertex.score = usize::MAX;
            vertex.parent = None;
        }
        prioqueue.push(vertex, -vertex.score);
    }

    // while Q is not empty:                      // The main loop
    //     u ← Q.extract_min()                    // Remove and return best vertex
    //     for each neighbor v of u:              // only v that are still in Q
    //         alt ← dist[u] + length(u, v)
    //         if alt < dist[v]
    //             dist[v] ← alt
    //             prev[v] ← u
    //             Q.decrease_priority(v, alt)
    while !prioqueue.is_empty() {
        let (i, j) = prioqueue.pop().unwrap();
        
        // for each neighbour still in prioqueue
        for di in 0..3 {
            for dj in 0..3 {
                let ii = i + di as usize;
                let jj = j + dj as usize;
                if ii == i && jj == j { continue; }
                if ii < 0 || nrows <= ii { continue; }
                if jj < 0 || ncols <= jj { continue; } 
                if prioqueue.get((ii, jj)).is_none() { continue; }
                let score = map[i][j].score + map[ii][jj].risk;
                if score < map[ii][jj].score {
                    map[ii][jj].score = score;
                    map[ii][jj].parent = (i, j);
                    prioqueue.change_priority(item: &Q, new_priority: P)
                }

            }
        }
    }

    // return dist, prev
    0
}

fn main() -> Result<(), Box<dyn Error>> {

    // Read input
    let input = fs::read_to_string("./data/test.txt")?;
    //let ncols = input.lines().next().unwrap().chars().count();
    //let nrows = input.lines().count();
    let mut graph: Vec<Rc<RefCell<Cell>>> = Vec::with_capacity(10000);
    input.lines().enumerate().for_each(|(i, line)| {
        line.trim()
            .chars()
            .map(|car| car.to_digit(10).unwrap())
            .enumerate()
            .for_each(|(j, risk)| {
                graph.push(Rc::new(RefCell::new(Cell::new(i, j, risk as usize))));
            })
    });

    // Get shortest path
    // todo

    Ok(())
}
